File: c++/expensetracker.cpp
#include <iostream> // For input/output operations (cin, cout)
#include <vector>   // For std::vector to store expenses
#include <string>   // For std::string to handle text data
#include <iomanip>  // For std::fixed and std::setprecision for formatting output
#include <map>      // For std::map to store category summaries
#include <algorithm> // For std::transform for case-insensitive comparison
#include <limits>   // For std::numeric_limits to clear input buffer
#include <cctype>    // For ::isdigit
#include <ctime>    // For tm struct, strptime, mktime

// Define a structure to represent an individual expense
// Using a struct makes all members public by default, which is suitable for a simple data container.
struct Expense {
    std::string date;        // Date of the expense (e.g., "MM-DD-YYYY")
    double amount;           // Amount of the expense
    std::string category;    // Category of the expense (e.g., "Food", "Transport")
    std::string description; // Description of the expense

    // Constructor to easily create Expense objects
    Expense(std::string d, double a, std::string c, std::string desc)
        : date(std::move(d)), amount(a), category(std::move(c)), description(std::move(desc)) {}
};

// Helper function to parse MM-DD-YYYY string to an integer YYYYMMDD for comparison.
// Returns -1 if the format is invalid or the date itself is invalid (e.g., Feb 30th).
long parseDateToInteger(const std::string& dateStr) {
    // Basic length check for "MM-DD-YYYY" format
    if (dateStr.length() != 10) {
        return -1;
    }

    struct tm tm_struct = {0}; // Initialize tm struct to all zeros

    // Use strptime to parse the date string into the tm struct.
    // %m: month as decimal number (01-12)
    // %d: day of month as decimal number (01-31)
    // %Y: year with century as decimal number
    // strptime returns a pointer to the character after the last character parsed, or NULL on error.
    char* parse_result = strptime(dateStr.c_str(), "%m-%d-%Y", &tm_struct);

    // Check if parsing was successful and the entire string was consumed (no extra characters)
    if (parse_result == NULL || *parse_result != '\0') {
        return -1; // Parsing failed or extra characters found in the string
    }

    // Convert tm struct to time_t to normalize values and validate the date.
    // mktime will adjust tm_mday, tm_mon, tm_year if they are out of range (e.g., if you pass Feb 30).
    // It returns (time_t)-1 on failure (e.g., completely invalid date that cannot be normalized).
    time_t time_val = mktime(&tm_struct);

    if (time_val == (time_t)-1) {
        // mktime failed, indicating an invalid date (e.g., non-existent date like Feb 30)
        return -1;
    }

    // After mktime, tm_struct contains normalized and valid date components.
    // We can now safely extract year, month, day and format to YYYYMMDD for comparison.
    // tm_year is years since 1900, tm_mon is 0-11
    int year = tm_struct.tm_year + 1900;
    int month = tm_struct.tm_mon + 1;
    int day = tm_struct.tm_mday;

    // Optional: Add a reasonable year range check if desired, though mktime handles much of the validation.
    if (year < 1900 || year > 2100) {
        return -1; // Date outside a reasonable application range
    }

    // Combine into a single long integer in YYYYMMDD format for easy chronological comparison
    return (long)year * 10000 + (long)month * 100 + day;
}


// Function to display a single expense
void displayExpense(const Expense& exp) {
    std::cout << std::fixed << std::setprecision(2); // Set precision for amount
    std::cout << "  Date: " << exp.date
              << ", Amount: $" << exp.amount
              << ", Category: " << exp.category
              << ", Description: " << exp.description << std::endl;
}

// Function to add a new expense
void addExpense(std::vector<Expense>& expenses) {
    std::string date, category, description;
    double amount;

    std::cout << "\n--- Add New Expense ---" << std::endl;
    std::cout << "Enter Date (MM-DD-YYYY): "; // Updated prompt
    // Input validation loop for date format
    while (true) {
        std::cin >> date;
        // Clear the input buffer after reading date string (important before getline)
        std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
        if (parseDateToInteger(date) != -1) {
            break; // Date format is valid, exit loop
        } else {
            std::cout << "Invalid date format or invalid date. Please use MM-DD-YYYY: "; // Updated error message
        }
    }

    std::cout << "Enter Amount: $";
    // Input validation for amount
    while (!(std::cin >> amount) || amount <= 0) {
        std::cout << "Invalid amount. Please enter a positive number: $";
        std::cin.clear(); // Clear error flags
        // Ignore remaining characters in the input buffer up to the newline
        std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
    }
    // Clear the input buffer after reading amount to prepare for getline
    std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');

    std::cout << "Enter Category (e.g., Food, Transport, Utilities): ";
    std::getline(std::cin, category); // Use getline to read category with spaces
    std::cout << "Enter Description: ";
    std::getline(std::cin, description); // Use getline to read description with spaces

    expenses.emplace_back(date, amount, category, description); // Add expense to vector
    std::cout << "Expense added successfully!" << std::endl;
}

// Function to view all expenses
void viewAllExpenses(const std::vector<Expense>& expenses) {
    std::cout << "\n--- All Expenses ---" << std::endl;
    if (expenses.empty()) {
        std::cout << "No expenses recorded yet." << std::endl;
        return;
    }
    for (const auto& exp : expenses) {
        displayExpense(exp);
    }
}

// Function to filter expenses by date range
void filterExpensesByDate(const std::vector<Expense>& expenses) {
    std::string startDateStr, endDateStr;
    long startDateInt, endDateInt;

    std::cout << "\n--- Filter Expenses by Date Range ---" << std::endl;
    std::cout << "Enter Start Date (MM-DD-YYYY): "; // Updated prompt
    // Input validation loop for start date format
    while (true) {
        std::cin >> startDateStr;
        std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n'); // Clear buffer
        startDateInt = parseDateToInteger(startDateStr);
        if (startDateInt != -1) {
            break;
        } else {
            std::cout << "Invalid date format or invalid date. Please use MM-DD-YYYY: "; // Updated error message
        }
    }

    std::cout << "Enter End Date (MM-DD-YYYY): "; // Updated prompt
    // Input validation loop for end date format
    while (true) {
        std::cin >> endDateStr;
        std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n'); // Clear buffer
        endDateInt = parseDateToInteger(endDateStr);
        if (endDateInt != -1) {
            break;
        } else {
            std::cout << "Invalid date format or invalid date. Please use MM-DD-YYYY: "; // Updated error message
        }
    }

    std::cout << "\nExpenses from " << startDateStr << " to " << endDateStr << ":" << std::endl;
    bool found = false;
    for (const auto& exp : expenses) {
        long expDateInt = parseDateToInteger(exp.date); // Convert expense date to integer

        // Compare using the integer representation of dates
        if (expDateInt != -1 && expDateInt >= startDateInt && expDateInt <= endDateInt) {
            displayExpense(exp);
            found = true;
        }
    }
    if (!found) {
        std::cout << "No expenses found in this date range." << std::endl;
    }
}

// Function to filter expenses by category
void filterExpensesByCategory(const std::vector<Expense>& expenses) {
    std::string categoryFilter;
    std::cout << "\n--- Filter Expenses by Category ---" << std::endl;
    std::cout << "Enter Category to filter by: ";
    std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n'); // Clear buffer before getline
    std::getline(std::cin, categoryFilter);

    std::cout << "\nExpenses in category '" << categoryFilter << "':" << std::endl;
    bool found = false;
    for (const auto& exp : expenses) {
        // Case-insensitive comparison for category
        std::string expCategoryLower = exp.category;
        std::string filterCategoryLower = categoryFilter;
        // Convert both strings to lowercase for comparison
        std::transform(expCategoryLower.begin(), expCategoryLower.end(), expCategoryLower.begin(), ::tolower);
        std::transform(filterCategoryLower.begin(), filterCategoryLower.end(), filterCategoryLower.begin(), ::tolower);

        if (expCategoryLower == filterCategoryLower) {
            displayExpense(exp);
            found = true;
        }
    }
    if (!found) {
        std::cout << "No expenses found for category '" << categoryFilter << "'." << std::endl;
    }
}

// Function to calculate and display summary of expenses
void showSummary(const std::vector<Expense>& expenses) {
    std::map<std::string, double> categoryTotals;
    double overallTotal = 0.0;

    for (const auto& exp : expenses) {
        categoryTotals[exp.category] += exp.amount;
        overallTotal += exp.amount;
    }

    std::cout << "\n--- Expense Summary ---" << std::endl;
    if (expenses.empty()) {
        std::cout << "No expenses recorded yet to summarize." << std::endl;
        return;
    }

    std::cout << "Total Expenses by Category:" << std::endl;
    std::cout << std::fixed << std::setprecision(2); // Set precision for amounts
    for (const auto& pair : categoryTotals) {
        std::cout << "  " << pair.first << ": $" << pair.second << std::endl;
    }

    std::cout << "\nOverall Total Expenses: $" << overallTotal << std::endl;
}

// Main function to run the application
int main() {
    std::vector<Expense> expenses; // Vector to store all expense objects
    int choice;

    do {
        std::cout << "\n--- Expense Tracker Menu ---" << std::endl;
        std::cout << "1. Add Expense" << std::endl;
        std::cout << "2. View All Expenses" << std::endl;
        std::cout << "3. Filter Expenses by Date Range" << std::endl;
        std::cout << "4. Filter Expenses by Category" << std::endl;
        std::cout << "5. Show Summary" << std::endl;
        std::cout << "6. Exit" << std::endl;
        std::cout << "Enter your choice: ";

        // Input validation for menu choice
        while (!(std::cin >> choice) || choice < 1 || choice > 6) {
            std::cout << "Invalid choice. Please enter a number between 1 and 6: ";
            std::cin.clear(); // Clear error flags
            std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n'); // Ignore remaining characters
        }

        switch (choice) {
            case 1:
                addExpense(expenses);
                break;
            case 2:
                viewAllExpenses(expenses);
                break;
            case 3:
                filterExpensesByDate(expenses);
                break;
            case 4:
                filterExpensesByCategory(expenses);
                break;
            case 5:
                showSummary(expenses);
                break;
            case 6:
                std::cout << "Exiting Expense Tracker. Goodbye!" << std::endl;
                break;
            default:
                // This case should ideally not be reached due to input validation
                std::cout << "An unexpected error occurred. Please try again." << std::endl;
                break;
        }
    } while (choice != 6); // Continue loop until user chooses to exit

    return 0; // Indicate successful execution
}

File: python/expense_tracker.py
import json
import os
from datetime import datetime, timedelta
from typing import List, Dict, Optional, Union
import uuid

class ExpenseTracker:
    """
    A comprehensive expense tracking application that demonstrates Python features:
    - Dictionary-based data storage
    - Dynamic typing capabilities
    - Datetime manipulation
    - JSON persistence
    """
    
    def __init__(self, data_file: str = "expenses.json"):
        """
        Initialize the expense tracker with optional data persistence.
        
        Args:
            data_file: Path to JSON file for data persistence
        """
        self.data_file = data_file
        # Primary data storage using dictionary - demonstrates Python's dict capabilities
        self.expenses: Dict[str, Dict] = {}
        # Categories for expense classification - demonstrates dynamic typing
        self.categories = {
            "food": "Food & Dining",
            "transport": "Transportation", 
            "utilities": "Utilities",
            "entertainment": "Entertainment",
            "healthcare": "Healthcare",
            "shopping": "Shopping",
            "education": "Education",
            "other": "Other"
        }
        # User options for expense tracking
        self.users = [
            "Bishal Thapa",
            "Aryan Shrestha", 
            "Mahesh Gaire",
            "Sajjad-IT-35",
            "SandeRestha",
            "Tushar Limbachiya"
        ]
        self.load_data()
    
    def add_expense(self, amountIn: Union[int, float], category: str,
                    description: str, user: str, date: Optional[str] = None) -> str:
        """
        Add a new expense to the tracker.
        Demonstrates Python's dynamic typing with Union types and optional parameters.
        
        Args:
            amountIn: Expense amount (int or float)
            category: Expense category
            description: Expense description
            user: User who made the expense
            date: Optional date string (YYYY-MM-DD), defaults to today
            
        Returns:
            str: Unique expense ID
        """
        # Generate unique ID
        expense_id = str(uuid.uuid4())[:8]
        
        # Handle date - demonstrate datetime manipulation
        if date is None:
            expense_date = datetime.now()
        else:
            try:
                expense_date = datetime.strptime(date, "%Y-%m-%d")
            except ValueError:
                raise ValueError("Date must be in YYYY-MM-DD format")
        
        # Validate category
        if category not in self.categories:
            available_categories = ", ".join(self.categories.keys())
            raise ValueError(f"Category must be one of: {available_categories}")
        
        # Validate user
        if user not in self.users:
            available_users = ", ".join(self.users)
            raise ValueError(f"User must be one of: {available_users}")
        
        # Validate amount - demonstrate dynamic typing
        if isinstance(amountIn, (int, float)) and amountIn > 0:
            expense_amount = float(amountIn)
        else:
            raise ValueError("Amount must be a positive number")
        
        # Create expense dictionary - demonstrates dict usage
        expense = {
            "id": expense_id,
            "amount": expense_amount,
            "category": category,
            "description": description.strip(),
            "user": user,
            "date": expense_date.strftime("%Y-%m-%d"),
            "timestamp": expense_date.isoformat(),
            "created_at": datetime.now().isoformat()
        }
        
        # Store in main dictionary
        self.expenses[expense_id] = expense
        self.save_data()
        
        return expense_id
    
    def view_all_expenses(self) -> List[Dict]:
        """
        Return all expenses sorted by date (newest first).
        
        Returns:
            List of expense dictionaries
        """
        expenses_list = list(self.expenses.values())
        # Sort by date using datetime parsing - demonstrates datetime usage
        expenses_list.sort(
            key=lambda x: datetime.fromisoformat(x["timestamp"]), 
            reverse=True
        )
        return expenses_list
    
    def filter_expenses_by_date(self, start_date: str, end_date: str) -> List[Dict]:
        """
        Filter expenses by date range.
        Demonstrates datetime manipulation and list comprehensions.
        
        Args:
            start_date: Start date string (YYYY-MM-DD)
            end_date: End date string (YYYY-MM-DD)
            
        Returns:
            List of filtered expense dictionaries
        """
        try:
            start = datetime.strptime(start_date, "%Y-%m-%d")
            end = datetime.strptime(end_date, "%Y-%m-%d")
        except ValueError:
            raise ValueError("Dates must be in YYYY-MM-DD format")
        
        if start > end:
            raise ValueError("Start date cannot be after end date")
        
        # Filter using list comprehension - demonstrates Python's concise syntax
        filtered = [
            expense for expense in self.expenses.values()
            if start <= datetime.fromisoformat(expense["timestamp"]) <= end
        ]
        
        # Sort by date
        filtered.sort(
            key=lambda x: datetime.fromisoformat(x["timestamp"]), 
            reverse=True
        )
        
        return filtered
    
    def filter_expenses_by_category(self, category: str) -> List[Dict]:
        """
        Filter expenses by category.
        
        Args:
            category: Category to filter by
            
        Returns:
            List of filtered expense dictionaries
        """
        if category not in self.categories:
            available_categories = ", ".join(self.categories.keys())
            raise ValueError(f"Category must be one of: {available_categories}")
        
        # Filter using list comprehension
        filtered = [
            expense for expense in self.expenses.values()
            if expense["category"] == category
        ]
        
        # Sort by date
        filtered.sort(
            key=lambda x: datetime.fromisoformat(x["timestamp"]), 
            reverse=True
        )
        
        return filtered
    
    def search_expenses(self, query: str) -> List[Dict]:
        """
        Search expenses by description text.
        Demonstrates string manipulation and filtering.
        
        Args:
            query: Search query string
            
        Returns:
            List of matching expense dictionaries
        """
        query_lower = query.lower().strip()
        
        # Search in descriptions using case-insensitive matching
        filtered = [
            expense for expense in self.expenses.values()
            if query_lower in expense["description"].lower()
        ]
        
        # Sort by date
        filtered.sort(
            key=lambda x: datetime.fromisoformat(x["timestamp"]), 
            reverse=True
        )
        
        return filtered
    
    def get_total_expenses(self) -> float:
        """
        Calculate total of all expenses.
        Demonstrates sum function with generator expression.
        
        Returns:
            Total amount of all expenses
        """
        return sum(expense["amount"] for expense in self.expenses.values())
    
    def get_expenses_by_category(self) -> Dict[str, Dict[str, Union[float, int]]]:
        """
        Get summary of expenses grouped by category.
        Demonstrates dictionary manipulation and aggregation.
        
        Returns:
            Dictionary with category summaries
        """
        summary = {}
        
        # Initialize all categories
        for category_key, category_name in self.categories.items():
            summary[category_key] = {
                "name": category_name,
                "total": 0.0,
                "count": 0
            }
        
        # Aggregate expenses by category
        for expense in self.expenses.values():
            category = expense["category"]
            summary[category]["total"] += expense["amount"]
            summary[category]["count"] += 1
        
        return summary
    
    def get_monthly_summary(self, year: int, month: int) -> Dict:
        """
        Get summary for a specific month.
        Demonstrates datetime manipulation and filtering.
        
        Args:
            year: Year (YYYY)
            month: Month (1-12)
            
        Returns:
            Dictionary with monthly summary
        """
        try:
            # Create start and end dates for the month
            start_date = datetime(year, month, 1)
            if month == 12:
                end_date = datetime(year + 1, 1, 1) - timedelta(days=1)
            else:
                end_date = datetime(year, month + 1, 1) - timedelta(days=1)
        except ValueError:
            raise ValueError("Invalid year or month")
        
        # Filter expenses for the month
        monthly_expenses = [
            expense for expense in self.expenses.values()
            if start_date <= datetime.fromisoformat(expense["timestamp"]) <= end_date
        ]
        
        # Calculate summary
        total_amount = sum(expense["amount"] for expense in monthly_expenses)
        expense_count = len(monthly_expenses)
        
        # Category breakdown
        category_totals = {}
        for expense in monthly_expenses:
            category = expense["category"]
            category_totals[category] = category_totals.get(category, 0) + expense["amount"]
        
        return {
            "month": f"{year}-{month:02d}",
            "total_amount": total_amount,
            "expense_count": expense_count,
            "category_breakdown": category_totals,
            "expenses": monthly_expenses
        }
    
    def delete_expense(self, expense_id: str) -> bool:
        """
        Delete an expense by ID.
        
        Args:
            expense_id: ID of expense to delete
            
        Returns:
            True if deleted, False if not found
        """
        if expense_id in self.expenses:
            del self.expenses[expense_id]
            self.save_data()
            return True
        return False
    
    def save_data(self) -> None:
        """
        Save expenses to JSON file.
        Demonstrates JSON serialization and file I/O.
        """
        try:
            with open(self.data_file, 'w') as f:
                json.dump(self.expenses, f, indent=2, default=str)
        except Exception as e:
            raise Exception(f"Could not save data to {self.data_file}: {e}")
    
    def load_data(self) -> None:
        """
        Load expenses from JSON file.
        Demonstrates JSON deserialization and error handling.
        """
        if os.path.exists(self.data_file):
            try:
                with open(self.data_file, 'r') as f:
                    self.expenses = json.load(f)
            except Exception as e:
                raise Exception(f"Could not load data from {self.data_file}: {e}")
        else:
            # File doesn't exist, start with empty expenses
            self.expenses = {}
    
    def _generate_unique_id(self) -> str:
        """Generate a unique ID that doesn't conflict with existing expenses."""
        import uuid
        while True:
            new_id = str(uuid.uuid4())[:8]
            if new_id not in self.expenses:
                return new_id
    
    def load_and_merge_data(self, filename: str) -> int:
        """
        Load expenses from a JSON file and merge with existing expenses.
        Returns the number of new expenses added.
        
        Args:
            filename: Path to the JSON file to load
            
        Returns:
            Number of new expenses that were merged
        """
        if not os.path.exists(filename):
            raise Exception(f"File {filename} does not exist")
        
        try:
            with open(filename, 'r') as f:
                file_data = json.load(f)
            
            # Handle different JSON formats
            if isinstance(file_data, dict):
                if 'expenses' in file_data:
                    # Export format with metadata
                    loaded_expenses = file_data['expenses']
                else:
                    # Direct expenses dictionary
                    loaded_expenses = file_data
            else:
                raise Exception("Invalid file format: expected JSON object")
            
            # Merge expenses, handling ID conflicts
            new_count = 0
            
            for expense_id, expense_data in loaded_expenses.items():
                if expense_id not in self.expenses:
                    # ID doesn't exist, add directly
                    self.expenses[expense_id] = expense_data
                    new_count += 1
                else:
                    # ID conflict, generate new unique ID
                    new_id = self._generate_unique_id()
                    expense_data['id'] = new_id  # Update the ID in the data
                    self.expenses[new_id] = expense_data
                    new_count += 1
            
            # Auto-save the merged data
            self.save_data()
            return new_count
            
        except json.JSONDecodeError as e:
            raise Exception(f"Invalid JSON file: {e}")
        except Exception as e:
            raise Exception(f"Could not load and merge data from {filename}: {e}")

def format_currency(amount: float) -> str:
    """Utility function to format currency."""
    return f"${amount:.2f}"

def format_expense_display(expense: Dict) -> str:
    """
    Format expense for display.
    Demonstrates string formatting and dictionary access.
    """
    return (
        f"ID: {expense['id']} | "
        f"Date: {expense['date']} | "
        f"Amount: {format_currency(expense['amount'])} | "
        f"Category: {expense['category'].title()} | "
        f"User: {expense.get('user', 'N/A')} | "
        f"Description: {expense['description']}"
    )

def display_expenses_table(expenses: List[Dict]) -> None:
    """
    Display expenses in a formatted table.
    Demonstrates string formatting and iteration.
    """
    if not expenses:
        print("No expenses found.")
        return
    
    print(f"\n{'='*130}")
    print(f"{'ID':<10} {'Date':<12} {'Amount':<12} {'Category':<15} {'User':<20} {'Description':<35}")
    print(f"{'='*130}")
    
    for expense in expenses:
        print(f"{expense['id']:<10} "
              f"{expense['date']:<12} "
              f"{format_currency(expense['amount']):<12} "
              f"{expense['category'].title():<15} "
              f"{expense.get('user', 'N/A'):<20} "
              f"{expense['description'][:33]:<35}")
    
    print(f"{'='*130}")
    total = sum(expense['amount'] for expense in expenses)
    print(f"Total: {format_currency(total)} ({len(expenses)} expenses)")

def main():
    """
    Main application loop with command-line interface.
    Demonstrates user interaction and menu system.
    """
    global amount
    tracker = ExpenseTracker()
    
    print("🏦 Welcome to the Expense Tracker Application!")
    print("This application demonstrates Python's dynamic typing, dictionary usage, and datetime manipulation.")
    
    while True:
        print(f"\n{'='*60}")
        print("EXPENSE TRACKER MENU")
        print(f"{'='*60}")
        print("1. Add Expense")
        print("2. View All Expenses")
        print("3. Filter by Date Range")
        print("4. Filter by Category")
        print("5. Search by Description")
        print("6. View Summary by Category")
        print("7. View Monthly Summary")
        print("8. Delete Expense")
        print("9. Exit")
        print(f"{'='*60}")
        
        try:
            choice = input("Select an option (1-9): ").strip()
            
            if choice == '1':
                # Add expense - demonstrates input validation and dynamic typing
                print("\n--- Add New Expense ---")
                amount_input = input("Enter amount: $").strip()
                try:
                    # Demonstrate dynamic typing - accept both int and float
                    amount = float(amount_input)
                except ValueError:
                    print("Invalid amount. Please enter a number.")
                    continue
                
                print("\nAvailable categories:")
                for key, name in tracker.categories.items():
                    print(f"  {key}: {name}")
                
                category = input("\nEnter category: ").strip().lower()
                description = input("Enter description: ").strip()
                
                print("\nAvailable users:")
                for i, user in enumerate(tracker.users, 1):
                    print(f"  {i}. {user}")
                
                user_input = input("\nEnter user number or name: ").strip()
                # Handle numeric input
                try:
                    user_index = int(user_input) - 1
                    if 0 <= user_index < len(tracker.users):
                        user = tracker.users[user_index]
                    else:
                        raise ValueError("Invalid user number")
                except ValueError:
                    # Try to match by name
                    user = user_input
                
                date_input = input("Enter date (YYYY-MM-DD) or press Enter for today: ").strip()
                date_value = date_input if date_input else None
                
                try:
                    expense_id = tracker.add_expense(amount, category, description, user, date_value)
                    print(f"Expense added successfully! ID: {expense_id}")
                except ValueError as e:
                    print(f"Error: {e}")
            
            elif choice == '2':
                # View all expenses
                print("\n--- All Expenses ---")
                expenses = tracker.view_all_expenses()
                display_expenses_table(expenses)
            
            elif choice == '3':
                # Filter by date range - demonstrates datetime manipulation
                print("\n--- Filter by Date Range ---")
                start_date = input("Enter start date (YYYY-MM-DD): ").strip()
                end_date = input("Enter end date (YYYY-MM-DD): ").strip()
                
                try:
                    expenses = tracker.filter_expenses_by_date(start_date, end_date)
                    print(f"\nExpenses from {start_date} to {end_date}:")
                    display_expenses_table(expenses)
                except ValueError as e:
                    print(f"Error: {e}")
            
            elif choice == '4':
                # Filter by category
                print("\n--- Filter by Category ---")
                print("Available categories:")
                for key, name in tracker.categories.items():
                    print(f"  {key}: {name}")
                
                category = input("\nEnter category: ").strip().lower()
                
                try:
                    expenses = tracker.filter_expenses_by_category(category)
                    print(f"\nExpenses in category '{category}':")
                    display_expenses_table(expenses)
                except ValueError as e:
                    print(f"Error: {e}")
            
            elif choice == '5':
                # Search by description
                print("\n--- Search by Description ---")
                query = input("Enter search query: ").strip()
                
                if query:
                    expenses = tracker.search_expenses(query)
                    print(f"\nExpenses containing '{query}':")
                    display_expenses_table(expenses)
                else:
                    print("Please enter a search query.")
            
            elif choice == '6':
                # View summary by category - demonstrates dictionary manipulation
                print("\n--- Expenses Summary by Category ---")
                summary = tracker.get_expenses_by_category()
                total_all = tracker.get_total_expenses()
                
                print(f"\n{'Category':<20} {'Total Amount':<15} {'Count':<8} {'Percentage':<12}")
                print(f"{'='*60}")
                
                for category_key, data in summary.items():
                    if data['count'] > 0:  # Only show categories with expenses
                        percentage = (data['total'] / total_all * 100) if total_all > 0 else 0
                        print(f"{data['name']:<20} "
                              f"{format_currency(data['total']):<15} "
                              f"{data['count']:<8} "
                              f"{percentage:.1f}%")
                
                print(f"{'='*60}")
                print(f"{'TOTAL':<20} {format_currency(total_all):<15} "
                      f"{sum(data['count'] for data in summary.values()):<8}")
            
            elif choice == '7':
                # Monthly summary - demonstrates datetime manipulation
                print("\n--- Monthly Summary ---")
                try:
                    year = int(input("Enter year (YYYY): ").strip())
                    month = int(input("Enter month (1-12): ").strip())
                    
                    summary = tracker.get_monthly_summary(year, month)
                    
                    print(f"\n--- Summary for {summary['month']} ---")
                    print(f"Total Amount: {format_currency(summary['total_amount'])}")
                    print(f"Number of Expenses: {summary['expense_count']}")
                    
                    if summary['category_breakdown']:
                        print(f"\nCategory Breakdown:")
                        for category, amount in summary['category_breakdown'].items():
                            category_name = tracker.categories.get(category, category)
                            print(f"  {category_name}: {format_currency(amount)}")
                    
                except ValueError as e:
                    print(f"Error: {e}")
            
            elif choice == '8':
                # Delete expense
                print("\n--- Delete Expense ---")
                expense_id = input("Enter expense ID to delete: ").strip()
                
                if tracker.delete_expense(expense_id):
                    print("Expense deleted successfully!")
                else:
                    print("Expense not found.")
            
            elif choice == '9':
                print("Thank you for using the Expense Tracker! ")
                break
            
            else:
                print("Invalid option. Please select 1-9.")
        
        except KeyboardInterrupt:
            print("\n\nGoodbye! ")
            break
        except Exception as e:
            print(f"An unexpected error occurred: {e}")

if __name__ == "__main__":
    main() 